Consul v1.12.2 - rpc error making call: i/o deadline reached

####  Overview of the Issue

After upgrade to Consul v1.12.2, we start seeing on the clients a lot of these errors:

[ERROR] agent.client: RPC failed to server: method=ConnectCA.Roots server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=Intention.Match server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"

Although Consul is throwing the erros,  everything seems to work ok.
We use Consul with ACL + TLS + Connect enabled + Nomad
Downgrading to Consul v1.11.3 (our previous version) the errors stop.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 2
	services = 2
build:
	prerelease =
	revision = 19041f20
	version = 1.12.2
	version_metadata =
consul:
	acl = enabled
	known_servers = 3
	server = false
runtime:
	arch = arm64
	cpu_count = 4
	goroutines = 117
	max_procs = 4
	os = linux
	version = go1.18.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 21
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 7
	member_time = 87554
	members = 16
	query_queue = 0
	query_time = 1
```

</details>

<details>
  <summary>Server info</summary>

```
agent:
	check_monitors = 0
	check_ttls = 1
	checks = 4
	services = 4
build:
	prerelease =
	revision = 19041f20
	version = 1.12.2
	version_metadata =
consul:
	acl = enabled
	bootstrap = false
	known_datacenters = 1
	leader = true
	leader_addr = xxx.xxx.xxx.xxx:8300
	server = true
raft:
	applied_index = 8198113
	commit_index = 8198113
	fsm_pending = 0
	last_contact = 0
	last_log_index = 8198113
	last_log_term = 189
	last_snapshot_index = 8188080
	last_snapshot_term = 189
	latest_configuration = [{Suffrage:Voter ID:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX Address:xxx.xxx.xxx.xxx:8300} {Suffrage:Voter ID:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX Address:xxx.xxx.xxx.xxx:8300} {Suffrage:Voter ID:XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX Address:xxx.xxx.xxx.xxx:8300}]
	latest_configuration_index = 0
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 189
runtime:
	arch = arm64
	cpu_count = 2
	goroutines = 378
	max_procs = 2
	os = linux
	version = go1.18.1
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 21
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 7
	member_time = 87556
	members = 16
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 123
	members = 3
	query_queue = 0
	query_time = 1
```

</details>

### Operating system and Environment details

AWS EC2, Debian11 / Ubuntu Server 20.04 (amd64/amr64)


### Log Fragments

[ERROR] agent.client: RPC failed to server: method=ConfigEntry.Get server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConnectCA.Roots server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=Intention.Match server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=Intention.Match server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.Get server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConfigEntry.ResolveServiceConfig server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"
[ERROR] agent.client: RPC failed to server: method=ConnectCA.Roots server=xxx.xxx.xxx.xxx:8300 error="rpc error making call: i/o deadline reached"

I also have those errors in version 1.13.1
```
│ 2022-08-12T20:59:07.862Z [ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=10.150.14.197:8300 error="rpc error making call: i/o deadline reached"                                                                             │
│ 2022-08-12T20:59:07.862Z [WARN]  agent.cache: handling error in Cache.Notify: cache-type=compiled-discovery-chain error="rpc error making call: i/o deadline reached" index=589824                                                                       │
│ 2022-08-12T20:59:08.203Z [ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=10.150.14.197:8300 error="rpc error making call: i/o deadline reached"                                                                             │
│ 2022-08-12T20:59:08.203Z [WARN]  agent.cache: handling error in Cache.Notify: cache-type=compiled-discovery-chain error="rpc error making call: i/o deadline reached" index=589824                                                                       │
│ 2022-08-12T20:59:09.818Z [ERROR] agent.client: RPC failed to server: method=Internal.PeeredUpstreams server=10.150.14.197:8300 error="rpc error making call: i/o deadline reached"                                                                       │
│ 2022-08-12T20:59:09.818Z [WARN]  agent.cache: handling error in Cache.Notify: cache-type=peered-upstreams error="rpc error making call: i/o deadline reached" index=1                                                                                    │
│ 2022-08-12T20:59:10.932Z [ERROR] agent.client: RPC failed to server: method=DiscoveryChain.Get server=10.150.14.197:8300 error="rpc error making call: i/o deadline reached"                                                                             │
│ 2022-08-12T20:59:10.933Z [WARN]  agent.cache: handling error in Cache.Notify: cache-type=compiled-discovery-chain error="rpc error making call: i/o deadline reached" index=589824
```

--- a/.changelog/14233.txt
+++ b/.changelog/14233.txt
@@ -1,0 +1,3 @@
+```release-note:bugfix
+rpc: Adds max jitter to client deadlines to prevent i/o deadline errors on blocking queries
+```

--- a/agent/consul/client_test.go
+++ b/agent/consul/client_test.go
@@ -893,8 +893,8 @@ func TestClient_RPC_Timeout(t *testing.T) {
 		}
 	})
 
-	// waiter will sleep for 50ms
-	require.NoError(t, s1.RegisterEndpoint("Wait", &waiter{duration: 50 * time.Millisecond}))
+	// waiter will sleep for 101ms which is 1ms more than the DefaultQueryTime
+	require.NoError(t, s1.RegisterEndpoint("Wait", &waiter{duration: 101 * time.Millisecond}))
 
 	// Requests with QueryOptions have a default timeout of RPCHoldTimeout (10ms)
 	// so we expect the RPC call to timeout.
@@ -903,7 +903,8 @@ func TestClient_RPC_Timeout(t *testing.T) {
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "rpc error making call: i/o deadline reached")
 
-	// Blocking requests have a longer timeout (100ms) so this should pass
+	// Blocking requests have a longer timeout (100ms) so this should pass since we
+	// add the maximum jitter which should be 16ms
 	out = struct{}{}
 	err = c1.RPC("Wait.Wait", &structs.NodeSpecificRequest{
 		QueryOptions: structs.QueryOptions{

--- a/agent/structs/structs.go
+++ b/agent/structs/structs.go
@@ -353,7 +353,7 @@ func (q QueryOptions) Timeout(rpcHoldTimeout, maxQueryTime, defaultQueryTime tim
 			q.MaxQueryTime = defaultQueryTime
 		}
 		// Timeout after maximum jitter has elapsed.
-		q.MaxQueryTime += lib.RandomStagger(q.MaxQueryTime / JitterFraction)
+		q.MaxQueryTime += q.MaxQueryTime / JitterFraction
 
 		return q.MaxQueryTime + rpcHoldTimeout
 	}

